File: pdu/python/hako_mavlink_msgs/pdu_conv_HakoHilStateQuaternion.py

import struct
from .pdu_pytype_HakoHilStateQuaternion import HakoHilStateQuaternion
from ..pdu_utils import *
from .. import binary_io

# dependencies for the generated Python class



def pdu_to_py_HakoHilStateQuaternion(binary_data: bytearray) -> HakoHilStateQuaternion:
    py_obj = HakoHilStateQuaternion()
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        raise ValueError("Invalid PDU binary data: MetaData not found or corrupted")
    binary_read_recursive_HakoHilStateQuaternion(meta, binary_data, py_obj, binary_io.PduMetaData.PDU_META_DATA_SIZE)
    return py_obj


def binary_read_recursive_HakoHilStateQuaternion(meta: binary_io.PduMetaData, binary_data: bytearray, py_obj: HakoHilStateQuaternion, base_off: int):
    # array_type: single 
    # data_type: primitive 
    # member_name: time_usec 
    # type_name: uint64 
    # offset: 0 size: 8 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 0, 8)
    py_obj.time_usec = binary_io.binTovalue("uint64", bin)
    
    # array_type: array 
    # data_type: primitive 
    # member_name: attitude_quaternion 
    # type_name: float32 
    # offset: 8 size: 16 
    # array_len: 4

    
    array_value = binary_io.readBinary(binary_data, base_off + 8, 16)
    py_obj.attitude_quaternion = binary_io.binToArrayValues("float32", array_value)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: rollspeed 
    # type_name: float32 
    # offset: 24 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 24, 4)
    py_obj.rollspeed = binary_io.binTovalue("float32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: pitchspeed 
    # type_name: float32 
    # offset: 28 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 28, 4)
    py_obj.pitchspeed = binary_io.binTovalue("float32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: yawspeed 
    # type_name: float32 
    # offset: 32 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 32, 4)
    py_obj.yawspeed = binary_io.binTovalue("float32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: lat 
    # type_name: int32 
    # offset: 36 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 36, 4)
    py_obj.lat = binary_io.binTovalue("int32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: lon 
    # type_name: int32 
    # offset: 40 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 40, 4)
    py_obj.lon = binary_io.binTovalue("int32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: alt 
    # type_name: int32 
    # offset: 44 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 44, 4)
    py_obj.alt = binary_io.binTovalue("int32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: vx 
    # type_name: int16 
    # offset: 48 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 48, 2)
    py_obj.vx = binary_io.binTovalue("int16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: vy 
    # type_name: int16 
    # offset: 50 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 50, 2)
    py_obj.vy = binary_io.binTovalue("int16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: vz 
    # type_name: int16 
    # offset: 52 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 52, 2)
    py_obj.vz = binary_io.binTovalue("int16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: ind_airspeed 
    # type_name: uint16 
    # offset: 54 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 54, 2)
    py_obj.ind_airspeed = binary_io.binTovalue("uint16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: true_airspeed 
    # type_name: uint16 
    # offset: 56 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 56, 2)
    py_obj.true_airspeed = binary_io.binTovalue("uint16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: xacc 
    # type_name: int16 
    # offset: 58 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 58, 2)
    py_obj.xacc = binary_io.binTovalue("int16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: yacc 
    # type_name: int16 
    # offset: 60 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 60, 2)
    py_obj.yacc = binary_io.binTovalue("int16", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: zacc 
    # type_name: int16 
    # offset: 62 size: 2 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 62, 2)
    py_obj.zacc = binary_io.binTovalue("int16", bin)
    
    return py_obj


def py_to_pdu_HakoHilStateQuaternion(py_obj: HakoHilStateQuaternion) -> bytearray:
    binary_data = bytearray()
    base_allocator = DynamicAllocator(False)
    bw_container = BinaryWriterContainer(binary_io.PduMetaData())
    binary_write_recursive_HakoHilStateQuaternion(0, bw_container, base_allocator, py_obj)

    # メタデータの設定
    total_size = base_allocator.size() + bw_container.heap_allocator.size() + binary_io.PduMetaData.PDU_META_DATA_SIZE
    bw_container.meta.total_size = total_size
    bw_container.meta.heap_off = binary_io.PduMetaData.PDU_META_DATA_SIZE + base_allocator.size()

    # binary_data のサイズを total_size に調整
    if len(binary_data) < total_size:
        binary_data.extend(bytearray(total_size - len(binary_data)))
    elif len(binary_data) > total_size:
        del binary_data[total_size:]

    # メタデータをバッファにコピー
    binary_io.writeBinary(binary_data, 0, bw_container.meta.to_bytes())

    # 基本データをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.base_off, base_allocator.to_array())

    # ヒープデータをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.heap_off, bw_container.heap_allocator.to_array())

    return binary_data

def binary_write_recursive_HakoHilStateQuaternion(parent_off: int, bw_container: BinaryWriterContainer, allocator, py_obj: HakoHilStateQuaternion):
    # array_type: single 
    # data_type: primitive 
    # member_name: time_usec 
    # type_name: uint64 
    # offset: 0 size: 8 
    # array_len: 1
    type = "uint64"
    off = 0

    
    bin = binary_io.typeTobin(type, py_obj.time_usec)
    bin = get_binary(type, bin, 8)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: array 
    # data_type: primitive 
    # member_name: attitude_quaternion 
    # type_name: float32 
    # offset: 8 size: 16 
    # array_len: 4
    type = "float32"
    off = 8

    
    elm_size =  16 
    array_size = int(4.0)
    one_elm_size = int(elm_size / array_size)
    binary = binary_io.typeTobin_array(type, py_obj.attitude_quaternion, one_elm_size)
    allocator.add(binary, expected_offset=(parent_off + off))
    
    # array_type: single 
    # data_type: primitive 
    # member_name: rollspeed 
    # type_name: float32 
    # offset: 24 size: 4 
    # array_len: 1
    type = "float32"
    off = 24

    
    bin = binary_io.typeTobin(type, py_obj.rollspeed)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: pitchspeed 
    # type_name: float32 
    # offset: 28 size: 4 
    # array_len: 1
    type = "float32"
    off = 28

    
    bin = binary_io.typeTobin(type, py_obj.pitchspeed)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: yawspeed 
    # type_name: float32 
    # offset: 32 size: 4 
    # array_len: 1
    type = "float32"
    off = 32

    
    bin = binary_io.typeTobin(type, py_obj.yawspeed)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: lat 
    # type_name: int32 
    # offset: 36 size: 4 
    # array_len: 1
    type = "int32"
    off = 36

    
    bin = binary_io.typeTobin(type, py_obj.lat)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: lon 
    # type_name: int32 
    # offset: 40 size: 4 
    # array_len: 1
    type = "int32"
    off = 40

    
    bin = binary_io.typeTobin(type, py_obj.lon)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: alt 
    # type_name: int32 
    # offset: 44 size: 4 
    # array_len: 1
    type = "int32"
    off = 44

    
    bin = binary_io.typeTobin(type, py_obj.alt)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: vx 
    # type_name: int16 
    # offset: 48 size: 2 
    # array_len: 1
    type = "int16"
    off = 48

    
    bin = binary_io.typeTobin(type, py_obj.vx)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: vy 
    # type_name: int16 
    # offset: 50 size: 2 
    # array_len: 1
    type = "int16"
    off = 50

    
    bin = binary_io.typeTobin(type, py_obj.vy)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: vz 
    # type_name: int16 
    # offset: 52 size: 2 
    # array_len: 1
    type = "int16"
    off = 52

    
    bin = binary_io.typeTobin(type, py_obj.vz)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: ind_airspeed 
    # type_name: uint16 
    # offset: 54 size: 2 
    # array_len: 1
    type = "uint16"
    off = 54

    
    bin = binary_io.typeTobin(type, py_obj.ind_airspeed)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: true_airspeed 
    # type_name: uint16 
    # offset: 56 size: 2 
    # array_len: 1
    type = "uint16"
    off = 56

    
    bin = binary_io.typeTobin(type, py_obj.true_airspeed)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: xacc 
    # type_name: int16 
    # offset: 58 size: 2 
    # array_len: 1
    type = "int16"
    off = 58

    
    bin = binary_io.typeTobin(type, py_obj.xacc)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: yacc 
    # type_name: int16 
    # offset: 60 size: 2 
    # array_len: 1
    type = "int16"
    off = 60

    
    bin = binary_io.typeTobin(type, py_obj.yacc)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: zacc 
    # type_name: int16 
    # offset: 62 size: 2 
    # array_len: 1
    type = "int16"
    off = 62

    
    bin = binary_io.typeTobin(type, py_obj.zacc)
    bin = get_binary(type, bin, 2)
    allocator.add(bin, expected_offset=parent_off + off)
    

if __name__ == "__main__":
    import sys
    import json

    def print_usage():
        print(f"Usage: python -m pdu.python.pdu_conv_HakoHilStateQuaternion <read|write> [args...]")
        print(f"  read <input_binary_file> <output_json_file>")
        print(f"  write <input_json_file> <output_binary_file>")

    if len(sys.argv) < 2:
        print_usage()
        sys.exit(1)

    command = sys.argv[1]

    if command == "read":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)
        
        binary_filepath = sys.argv[2]
        output_json_filepath = sys.argv[3]

        with open(binary_filepath, "rb") as f:
            binary_data = bytearray(f.read())
        
        py_obj = pdu_to_py_HakoHilStateQuaternion(binary_data)
        
        with open(output_json_filepath, "w") as f:
            f.write(py_obj.to_json())

    elif command == "write":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)

        input_json_filepath = sys.argv[2]
        output_binary_filepath = sys.argv[3]

        with open(input_json_filepath, "r") as f:
            json_str = f.read()
        
        py_obj = HakoHilStateQuaternion.from_json(json_str)
        
        binary_data = py_to_pdu_HakoHilStateQuaternion(py_obj)

        with open(output_binary_filepath, "wb") as f:
            f.write(binary_data)

    else:
        print(f"Unknown command: {command}")
        print_usage()
        sys.exit(1)

File: pdu/javascript/hako_mavlink_msgs/pdu_conv_HakoHilStateQuaternion.js
import * as PduUtils from '../pdu_utils.js';
import { HakoHilStateQuaternion } from './pdu_jstype_HakoHilStateQuaternion.js';


/**
 * Deserializes a binary PDU into a HakoHilStateQuaternion object.
 * @param {ArrayBuffer} binary_data
 * @returns { HakoHilStateQuaternion }
 */
export function pduToJs_HakoHilStateQuaternion(binary_data) {
    const js_obj = new HakoHilStateQuaternion();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_HakoHilStateQuaternion(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_HakoHilStateQuaternion(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: time_usec, type: uint64 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 0, 8);
        js_obj.time_usec = PduUtils.binToValue("uint64", bin);
    }
    
    // member: attitude_quaternion, type: float32 (primitive)

    
    {
        const array_bin = PduUtils.readBinary(binary_data, base_off + 8, 16);
        js_obj.attitude_quaternion = PduUtils.binToArrayValues("float32", array_bin, 4);
    }
    
    // member: rollspeed, type: float32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 24, 4);
        js_obj.rollspeed = PduUtils.binToValue("float32", bin);
    }
    
    // member: pitchspeed, type: float32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 28, 4);
        js_obj.pitchspeed = PduUtils.binToValue("float32", bin);
    }
    
    // member: yawspeed, type: float32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 32, 4);
        js_obj.yawspeed = PduUtils.binToValue("float32", bin);
    }
    
    // member: lat, type: int32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 36, 4);
        js_obj.lat = PduUtils.binToValue("int32", bin);
    }
    
    // member: lon, type: int32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 40, 4);
        js_obj.lon = PduUtils.binToValue("int32", bin);
    }
    
    // member: alt, type: int32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 44, 4);
        js_obj.alt = PduUtils.binToValue("int32", bin);
    }
    
    // member: vx, type: int16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 48, 2);
        js_obj.vx = PduUtils.binToValue("int16", bin);
    }
    
    // member: vy, type: int16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 50, 2);
        js_obj.vy = PduUtils.binToValue("int16", bin);
    }
    
    // member: vz, type: int16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 52, 2);
        js_obj.vz = PduUtils.binToValue("int16", bin);
    }
    
    // member: ind_airspeed, type: uint16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 54, 2);
        js_obj.ind_airspeed = PduUtils.binToValue("uint16", bin);
    }
    
    // member: true_airspeed, type: uint16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 56, 2);
        js_obj.true_airspeed = PduUtils.binToValue("uint16", bin);
    }
    
    // member: xacc, type: int16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 58, 2);
        js_obj.xacc = PduUtils.binToValue("int16", bin);
    }
    
    // member: yacc, type: int16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 60, 2);
        js_obj.yacc = PduUtils.binToValue("int16", bin);
    }
    
    // member: zacc, type: int16 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 62, 2);
        js_obj.zacc = PduUtils.binToValue("int16", bin);
    }
    
    return js_obj;
}

/**
 * Serializes a HakoHilStateQuaternion object into a binary PDU.
 * @param { HakoHilStateQuaternion } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_HakoHilStateQuaternion(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_HakoHilStateQuaternion(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_HakoHilStateQuaternion(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: time_usec, type: uint64 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint64", js_obj.time_usec, 8);
        allocator.add(bin, parent_off + 0);
    }
    
    // member: attitude_quaternion, type: float32 (primitive)

    
    {
        const buffer = PduUtils.typesToBin("float32", js_obj.attitude_quaternion);
        allocator.add(buffer, parent_off + 8);
    }
    
    // member: rollspeed, type: float32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("float32", js_obj.rollspeed, 4);
        allocator.add(bin, parent_off + 24);
    }
    
    // member: pitchspeed, type: float32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("float32", js_obj.pitchspeed, 4);
        allocator.add(bin, parent_off + 28);
    }
    
    // member: yawspeed, type: float32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("float32", js_obj.yawspeed, 4);
        allocator.add(bin, parent_off + 32);
    }
    
    // member: lat, type: int32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int32", js_obj.lat, 4);
        allocator.add(bin, parent_off + 36);
    }
    
    // member: lon, type: int32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int32", js_obj.lon, 4);
        allocator.add(bin, parent_off + 40);
    }
    
    // member: alt, type: int32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int32", js_obj.alt, 4);
        allocator.add(bin, parent_off + 44);
    }
    
    // member: vx, type: int16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int16", js_obj.vx, 2);
        allocator.add(bin, parent_off + 48);
    }
    
    // member: vy, type: int16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int16", js_obj.vy, 2);
        allocator.add(bin, parent_off + 50);
    }
    
    // member: vz, type: int16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int16", js_obj.vz, 2);
        allocator.add(bin, parent_off + 52);
    }
    
    // member: ind_airspeed, type: uint16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint16", js_obj.ind_airspeed, 2);
        allocator.add(bin, parent_off + 54);
    }
    
    // member: true_airspeed, type: uint16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint16", js_obj.true_airspeed, 2);
        allocator.add(bin, parent_off + 56);
    }
    
    // member: xacc, type: int16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int16", js_obj.xacc, 2);
        allocator.add(bin, parent_off + 58);
    }
    
    // member: yacc, type: int16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int16", js_obj.yacc, 2);
        allocator.add(bin, parent_off + 60);
    }
    
    // member: zacc, type: int16 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int16", js_obj.zacc, 2);
        allocator.add(bin, parent_off + 62);
    }
    
}
